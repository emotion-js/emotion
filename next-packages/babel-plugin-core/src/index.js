// @flow
import cssMacro, { transformCssCallExpression } from './css-macro'
import styledMacro from './styled-macro'
import { createPlugin } from './create-plugin'
import { addDefault } from '@babel/helper-module-imports'
import { getSourceMap } from '@emotion/babel-utils'

export const macros = {
  css: cssMacro,
  styled: styledMacro
}

export const appendStringToExpressions = (
  expressions: Array<*>,
  string: string,
  t: *
) => {
  if (!string) {
    return expressions
  }
  if (t.isStringLiteral(expressions[expressions.length - 1])) {
    expressions[expressions.length - 1].value += string
  } else {
    expressions.push(t.stringLiteral(string))
  }
  return expressions
}

let emotionCoreMacroThatsNotARealMacro = ({ references, state, babel }) => {
  Object.keys(references).forEach(refKey => {
    if (refKey === 'css') {
      references[refKey].forEach(path => {
        transformCssCallExpression({ babel, state, path: path.parentPath })
      })
    }
  })
}
emotionCoreMacroThatsNotARealMacro.keepImport = true

export default createPlugin(
  {
    '@emotion/css': cssMacro,
    '@emotion/styled': styledMacro,
    '@emotion/core': emotionCoreMacroThatsNotARealMacro
  },
  babel => {
    let t = babel.types
    return {
      Program(path: *, state: *) {
        if (state.opts.jsx === undefined) {
          for (const node of path.node.body) {
            if (
              t.isImportDeclaration(node) &&
              node.source.value === '@emotion/core'
            ) {
              state.transformCssProp = true
              break
            }
          }
        } else {
          state.transformCssProp = state.opts.jsx
        }
        if (state.opts.sourceMap) {
          state.emotionSourceMap = true
        }
      },
      JSXAttribute(path: *, state: *) {
        if (path.node.name.name !== 'css' && state.transformCssProp) {
          return
        }

        if (
          t.isJSXExpressionContainer(path.node.value) &&
          (t.isObjectExpression(path.node.value.expression) ||
            t.isArrayExpression(path.node.value.expression))
        ) {
          let expressionPath = path.get('value.expression')
          if (expressionPath.isPure()) {
            if (!state.cssIdentifier) {
              state.cssIdentifier = addDefault(path, '@emotion/css', {
                nameHint: 'css'
              })
            }
            expressionPath.replaceWith(
              t.callExpression(
                t.cloneDeep(state.cssIdentifier),
                [
                  path.node.value.expression,
                  state.emotionSourceMap &&
                    path.node.loc !== undefined &&
                    t.stringLiteral(getSourceMap(path.node.loc.start, state))
                ].filter(Boolean)
              )
            )
            transformCssCallExpression({
              babel,
              state,
              path: expressionPath
            })
          }
        }
      }
    }
  }
)
