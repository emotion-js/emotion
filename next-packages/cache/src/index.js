// @flow
import { StyleSheet } from '@emotion/sheet'
import { isBrowser, type CSSContextType } from '@emotion/utils'
import Stylis from '@emotion/stylis'
import ruleSheetPlugin from './rule-sheet'
import type { StylisPlugin } from './types'

export type PrefixOption =
  | boolean
  | ((key: string, value: string, context: 1 | 2 | 3) => boolean)

type StylisPlugins = StylisPlugin[] | StylisPlugin

export type Options = {
  nonce?: string,
  stylisPlugins?: StylisPlugins,
  prefix?: PrefixOption,
  key?: string,
  container?: HTMLElement
}

let createCache = (options?: Options): CSSContextType => {
  if (options === undefined) options = {}
  let key = options.key || 'css'
  let stylisOptions

  if (options.prefix !== undefined) {
    stylisOptions = {
      prefix: options.prefix
    }
  }

  let stylis = new Stylis(stylisOptions)

  stylis.use(options.stylisPlugins)(ruleSheetPlugin)

  if (process.env.NODE_ENV !== 'production') {
    // $FlowFixMe
    if (/[^a-z-]/.test(key)) {
      throw new Error(
        `Emotion key must only contain lower case alphabetical characters and - but "${key}" was passed`
      )
    }
    let sourceMapRegEx = /\/\*#\ssourceMappingURL=data:application\/json;\S+\s+\*\//
    let currentSourceMap
    stylis.use((context, content, selectors) => {
      switch (context) {
        case -1: {
          let result = sourceMapRegEx.exec(content)
          if (result) {
            currentSourceMap = result[0]
          }
          break
        }
        case 2: {
          for (let i = 0, len = selectors.length; len > i; i++) {
            // :last-child isn't included here since it's safe
            // because a style element will never be the last element
            let match = selectors[i].match(/:(first|nth|nth-last)-child/)
            if (match !== null) {
              console.error(
                `The pseudo class "${
                  match[1]
                }" is potentially unsafe when doing server-side rendering. Try changing it to "${
                  match[1]
                }-of-type"`
              )
            }
          }
          break
        }
        case -2: {
          if (currentSourceMap) {
            content.forEach((rule, i) => {
              content[i] = rule + currentSourceMap
            })

            currentSourceMap = ''
          }
        }
      }
    })
  }
  let inserted = {}

  if (isBrowser) {
    // $FlowFixMe
    let parent: HTMLElement = options.container || document.head

    const nodes = document.querySelectorAll(`style[data-emotion-${key}]`)

    Array.prototype.forEach.call(nodes, (node: HTMLStyleElement) => {
      const attrib = node.getAttribute(`data-emotion-${key}`)
      // $FlowFixMe
      attrib.split(' ').forEach(id => {
        inserted[id] = true
      })
      if (node.parentNode !== parent) {
        parent.appendChild(node)
      }
    })
  }

  const context: CSSContextType = {
    stylis,
    key,
    sheet: new StyleSheet({
      key,
      container: options.container,
      nonce: options.nonce
    }),
    inserted,
    registered: {},
    theme: {}
  }
  return context
}

export default createCache
