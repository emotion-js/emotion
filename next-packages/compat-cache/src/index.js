// @flow
import createCache from '@emotion/cache'

let createCompatCache = (emotion: *) => {
  // $FlowFixMe
  let cache = createCache(emotion.sheet.opts)
  cache.inserted = emotion.caches.inserted
  cache.registered = emotion.caches.registered
  cache.compat = true
  return cache
}

export default createCompatCache
