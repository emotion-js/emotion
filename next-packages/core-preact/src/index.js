// an empty file since this is replaced with @emotion/core but with react aliased to our react mock
