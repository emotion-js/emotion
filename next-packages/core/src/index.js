// @flow
export { withCSSContext, Provider } from './context'
export { jsx } from './jsx'
export { Global } from './global'
export { keyframes } from './keyframes'
export { default as css } from '@emotion/css'
