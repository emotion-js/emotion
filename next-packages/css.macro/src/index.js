// @flow

module.exports = require('@emotion/babel-plugin-core').macros.css
