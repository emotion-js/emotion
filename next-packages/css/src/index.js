// @flow
import type { Interpolation, ScopedInsertableStyles } from '@emotion/utils'
import { serializeStyles } from '@emotion/serialize'

let fakeRegisteredCache = {}

function css(...args: Array<Interpolation>): ScopedInsertableStyles {
  return serializeStyles(fakeRegisteredCache, args)
}

export default css
