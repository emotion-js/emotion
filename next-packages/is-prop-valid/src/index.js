import memoize from '@emotion/memoize'

declare var codegen: { require: string => RegExp }

const reactPropsRegex = codegen.require('./props')

export default memoize(RegExp.prototype.test.bind(reactPropsRegex))
