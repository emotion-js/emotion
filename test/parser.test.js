import { parseCSS } from '../src/parser'

const basicCSS = `.thing {
  display: flex;
  justify-content: center;
  width: var(--css-hash-0);
}`

const fancyCSS = `.some-selector {
  display: flex;
  justify-content: center;
  width: var(--css-hash-0);
  &:hover {
    background-color: green;
  }
  @media (max-width: 500px) {
    height: var(--css-hash-1);
    position: fixed;
  }
  @media print {
    display: none;
  }
  &::before {
    color: blue;
    width: 20px;
    height: 20px;
    content: 'pseudo';
  }
}`

const staticCSS = `.thing {
  display: block;
  height: 50px;
  width: 30px;
}`

describe('parser', () => {
  test('basic', () => {
    expect(parseCSS(basicCSS)).toMatchSnapshot()
  })
  test('fancy', () => {
    expect(parseCSS(fancyCSS)).toMatchSnapshot()
  })
  test('static', () => {
    expect(parseCSS(staticCSS, true)).toMatchSnapshot()
  })
})
