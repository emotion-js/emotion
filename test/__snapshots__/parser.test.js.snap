// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser basic 1`] = `
Object {
  "composes": 0,
  "hasCssFunction": false,
  "hasOtherMatch": false,
  "hasVar": false,
  "rules": Array [
    ".thing {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  -webkit-box-pack: center;
  justify-content: center;
  width: var(--css-hash-0);
}",
  ],
}
`;

exports[`parser fancy 1`] = `
Object {
  "composes": 0,
  "hasCssFunction": false,
  "hasOtherMatch": false,
  "hasVar": false,
  "rules": Array [
    ".some-selector {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  -webkit-box-pack: center;
  justify-content: center;
  width: var(--css-hash-0);
}",
    ".some-selector:hover {
  background-color: green;
}",
    "@media (max-width: 500px) {
  .some-selector {
    height: var(--css-hash-1);
    position: fixed;
  }
}",
    "@media print {
  .some-selector {
    display: none;
  }
}",
    ".some-selector::before {
  color: blue;
  width: 20px;
  height: 20px;
  content: 'pseudo'
}",
  ],
}
`;

exports[`parser static 1`] = `
Object {
  "composes": 0,
  "hasCssFunction": false,
  "hasOtherMatch": false,
  "hasVar": false,
  "rules": Array [
    ".thing {
  display: block;
  height: 50px;
  width: 30px
}",
  ],
}
`;
